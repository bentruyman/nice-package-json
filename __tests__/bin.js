"use strict";

const { execFile } = require("child_process");
const { copyFileSync, readFileSync, statSync } = require("fs");
const { join } = require("path");

const tmp = require("tmp");

const ROOT = join(__dirname, "..");
const BIN = join(ROOT, "bin.js");
const FIXTURES = join(__dirname, "fixtures");

function exec(options = {}, args = []) {
  return new Promise((resolve) => {
    execFile("node", [BIN, ...args], options, (err, stdout, stderr) => {
      resolve({ err, stdout, stderr });
    });
  });
}

describe("$ nice-package-json", () => {
  it("prints a formatted package.json from the current directory", async () => {
    const dir = tmp.dirSync();

    copyFileSync(
      join(FIXTURES, "input/full.json"),
      join(dir.name, "package.json")
    );

    const { err, stdout, stderr } = await exec({ cwd: dir.name });
    expect(err).toBeNull();
    expect(stdout).toMatchSnapshot();
    expect(stderr).toBe("");
  });

  it("shows an error when no package.json file is found", async () => {
    const dir = tmp.dirSync();

    const { err, stdout, stderr } = await exec({ cwd: __dirname });
    expect(err).not.toBeNull();
    expect(stdout).toBe("");
    expect(stderr).toMatchSnapshot();
  });

  describe(" --write", () => {
    it("overwrites an existing package.json if changes are made", async () => {
      const dir = tmp.dirSync();
      const inputFile = join(FIXTURES, "input/full.json");
      const expectedFile = join(FIXTURES, "expected/full.json");
      const targetPkg = join(dir.name, "package.json");

      copyFileSync(inputFile, targetPkg);

      expect(readFileSync(targetPkg)).toEqual(readFileSync(inputFile));
      await exec({ cwd: dir.name }, ["--write"]);
      expect(readFileSync(targetPkg)).toEqual(readFileSync(expectedFile));
    });

    it("doesn't touch an existing package.json if no changes are made", async () => {
      const dir = tmp.dirSync();
      const inputFile = join(FIXTURES, "expected/full.json");
      const targetPkg = join(dir.name, "package.json");

      copyFileSync(inputFile, targetPkg);

      const beforeModifiedTime = statSync(targetPkg).mtimeMs;
      await exec({ cwd: dir.name }, ["--write"]);
      const afterModifiedTime = statSync(targetPkg).mtimeMs;

      expect(beforeModifiedTime).toBe(afterModifiedTime);
    });
  });
});

describe("$ nice-package-json /path/to/dir", () => {
  it("prints a formatted package.json from the specified directory", async () => {
    const dir = tmp.dirSync();

    copyFileSync(
      join(FIXTURES, "input/full.json"),
      join(dir.name, "package.json")
    );
    const { err, stdout, stderr } = await exec({ cwd: __dirname }, [dir.name]);
    expect(err).toBeNull();
    expect(stdout).toMatchSnapshot();
    expect(stderr).toBe("");
  });

  it("shows an error when no package.json file is found", async () => {
    const { err, stdout, stderr } = await exec({ cwd: __dirname }, [
      "/missing/dir",
    ]);
    expect(err.message).toMatch("no package.json file found");
    expect(stdout).toMatchSnapshot();
    expect(stderr).toMatch("no package.json file found");
  });
});

describe("$ nice-package-json /path/to/package.json", () => {
  it("prints a formatted package.json using the specified file", async () => {
    const dir = tmp.dirSync();

    copyFileSync(
      join(FIXTURES, "input/full.json"),
      join(dir.name, "package.json")
    );
    const { err, stdout, stderr } = await exec({ cwd: dir.name }, [
      "./package.json",
    ]);
    expect(err).toBeNull();
    expect(stdout).toMatchSnapshot();
    expect(stderr).toBe("");
  });

  it("shows an error if the file is not valid JSON", async () => {
    const { err, stdout, stderr } = await exec({ cwd: __dirname }, [
      "/missing/dir/package.json",
    ]);
    expect(err.message).toMatch("no package.json file found");
    expect(stdout).toMatchSnapshot();
    expect(stderr).toMatch("no package.json file found");
  });
});
