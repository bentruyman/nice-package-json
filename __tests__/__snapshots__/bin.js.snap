// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$ nice-package-json /path/to/dir prints a formatted package.json from the specified directory 1`] = `
"{
  \\"name\\": \\"full\\",
  \\"private\\": \\"true\\",
  \\"version\\": \\"1.2.3\\",
  \\"description\\": \\"Package description\\",
  \\"keywords\\": [
    \\"a\\",
    \\"b\\",
    \\"c\\"
  ],
  \\"engines\\": {
    \\"node\\": \\"12.0.0\\",
    \\"npm\\": \\"6.0.0\\"
  },
  \\"type\\": \\"module\\",
  \\"main\\": \\"dist/index.js\\",
  \\"module\\": \\"dist/index.mjs\\",
  \\"svelte\\": \\"src/index.js\\",
  \\"bin\\": \\"bin.js\\",
  \\"files\\": [
    \\"src\\",
    \\"test\\"
  ],
  \\"homepage\\": \\"https://host.com/test/repo\\",
  \\"bugs\\": \\"https://host.com/test/repo/issues\\",
  \\"repository\\": {
    \\"type\\": \\"git\\",
    \\"url\\": \\"git@host.com:test/repo.git\\"
  },
  \\"author\\": \\"Barney Rubble <b@rubble.com>\\",
  \\"contributors\\": [
    {
      \\"name\\": \\"Jane Smith\\",
      \\"email\\": \\"jane.smith@host.com\\"
    },
    \\"Jane Smith <jane.smith@host.com>\\",
    {
      \\"name\\": \\"John Smith\\",
      \\"email\\": \\"john.smith@host.com\\"
    },
    \\"John Smith <john.smith@host.com>\\"
  ],
  \\"license\\": \\"MIT\\",
  \\"scripts\\": {
    \\"build\\": \\"./build.sh\\",
    \\"lint\\": \\"./lint.sh\\",
    \\"test\\": \\"./test.sh\\"
  },
  \\"dependencies\\": {
    \\"express\\": \\"^1.0.0\\",
    \\"lodash\\": \\"^2.0.0\\"
  },
  \\"devDependencies\\": {
    \\"eslint\\": \\"^1.0.0\\",
    \\"jest\\": \\"^2.0.0\\"
  },
  \\"bundledDependencies\\": {
    \\"express\\": \\"^1.0.0\\",
    \\"lodash\\": \\"^2.0.0\\"
  },
  \\"optionalDependencies\\": {
    \\"bluebird\\": \\"^1.0.0\\",
    \\"fibers\\": \\"^2.0.0\\"
  },
  \\"peerDependencies\\": {
    \\"vue\\": \\"^1.0.0\\",
    \\"vue-template-compiler\\": \\"^2.0.0\\"
  },
  \\"publishConfig\\": {
    \\"access\\": \\"public\\",
    \\"registry\\": \\"https://internal.host.com/registry\\"
  },
  \\"eslintConfig\\": {
    \\"extends\\": \\"standard\\"
  },
  \\"jest\\": {
    \\"bail\\": 1,
    \\"verbose\\": true
  }
}
"
`;

exports[`$ nice-package-json /path/to/dir shows an error when no package.json file is found 1`] = `""`;

exports[`$ nice-package-json /path/to/package.json prints a formatted package.json using the specified file 1`] = `
"{
  \\"name\\": \\"full\\",
  \\"private\\": \\"true\\",
  \\"version\\": \\"1.2.3\\",
  \\"description\\": \\"Package description\\",
  \\"keywords\\": [
    \\"a\\",
    \\"b\\",
    \\"c\\"
  ],
  \\"engines\\": {
    \\"node\\": \\"12.0.0\\",
    \\"npm\\": \\"6.0.0\\"
  },
  \\"type\\": \\"module\\",
  \\"main\\": \\"dist/index.js\\",
  \\"module\\": \\"dist/index.mjs\\",
  \\"svelte\\": \\"src/index.js\\",
  \\"bin\\": \\"bin.js\\",
  \\"files\\": [
    \\"src\\",
    \\"test\\"
  ],
  \\"homepage\\": \\"https://host.com/test/repo\\",
  \\"bugs\\": \\"https://host.com/test/repo/issues\\",
  \\"repository\\": {
    \\"type\\": \\"git\\",
    \\"url\\": \\"git@host.com:test/repo.git\\"
  },
  \\"author\\": \\"Barney Rubble <b@rubble.com>\\",
  \\"contributors\\": [
    {
      \\"name\\": \\"Jane Smith\\",
      \\"email\\": \\"jane.smith@host.com\\"
    },
    \\"Jane Smith <jane.smith@host.com>\\",
    {
      \\"name\\": \\"John Smith\\",
      \\"email\\": \\"john.smith@host.com\\"
    },
    \\"John Smith <john.smith@host.com>\\"
  ],
  \\"license\\": \\"MIT\\",
  \\"scripts\\": {
    \\"build\\": \\"./build.sh\\",
    \\"lint\\": \\"./lint.sh\\",
    \\"test\\": \\"./test.sh\\"
  },
  \\"dependencies\\": {
    \\"express\\": \\"^1.0.0\\",
    \\"lodash\\": \\"^2.0.0\\"
  },
  \\"devDependencies\\": {
    \\"eslint\\": \\"^1.0.0\\",
    \\"jest\\": \\"^2.0.0\\"
  },
  \\"bundledDependencies\\": {
    \\"express\\": \\"^1.0.0\\",
    \\"lodash\\": \\"^2.0.0\\"
  },
  \\"optionalDependencies\\": {
    \\"bluebird\\": \\"^1.0.0\\",
    \\"fibers\\": \\"^2.0.0\\"
  },
  \\"peerDependencies\\": {
    \\"vue\\": \\"^1.0.0\\",
    \\"vue-template-compiler\\": \\"^2.0.0\\"
  },
  \\"publishConfig\\": {
    \\"access\\": \\"public\\",
    \\"registry\\": \\"https://internal.host.com/registry\\"
  },
  \\"eslintConfig\\": {
    \\"extends\\": \\"standard\\"
  },
  \\"jest\\": {
    \\"bail\\": 1,
    \\"verbose\\": true
  }
}
"
`;

exports[`$ nice-package-json /path/to/package.json shows an error if the file is not valid JSON 1`] = `""`;

exports[`$ nice-package-json {foo,bar}/package.json prints formatted package.json files using the specified file 1`] = `
"{
  \\"name\\": \\"full\\",
  \\"private\\": \\"true\\",
  \\"version\\": \\"1.2.3\\",
  \\"description\\": \\"Package description\\",
  \\"keywords\\": [
    \\"a\\",
    \\"b\\",
    \\"c\\"
  ],
  \\"engines\\": {
    \\"node\\": \\"12.0.0\\",
    \\"npm\\": \\"6.0.0\\"
  },
  \\"type\\": \\"module\\",
  \\"main\\": \\"dist/index.js\\",
  \\"module\\": \\"dist/index.mjs\\",
  \\"svelte\\": \\"src/index.js\\",
  \\"bin\\": \\"bin.js\\",
  \\"files\\": [
    \\"src\\",
    \\"test\\"
  ],
  \\"homepage\\": \\"https://host.com/test/repo\\",
  \\"bugs\\": \\"https://host.com/test/repo/issues\\",
  \\"repository\\": {
    \\"type\\": \\"git\\",
    \\"url\\": \\"git@host.com:test/repo.git\\"
  },
  \\"author\\": \\"Barney Rubble <b@rubble.com>\\",
  \\"contributors\\": [
    {
      \\"name\\": \\"Jane Smith\\",
      \\"email\\": \\"jane.smith@host.com\\"
    },
    \\"Jane Smith <jane.smith@host.com>\\",
    {
      \\"name\\": \\"John Smith\\",
      \\"email\\": \\"john.smith@host.com\\"
    },
    \\"John Smith <john.smith@host.com>\\"
  ],
  \\"license\\": \\"MIT\\",
  \\"scripts\\": {
    \\"build\\": \\"./build.sh\\",
    \\"lint\\": \\"./lint.sh\\",
    \\"test\\": \\"./test.sh\\"
  },
  \\"dependencies\\": {
    \\"express\\": \\"^1.0.0\\",
    \\"lodash\\": \\"^2.0.0\\"
  },
  \\"devDependencies\\": {
    \\"eslint\\": \\"^1.0.0\\",
    \\"jest\\": \\"^2.0.0\\"
  },
  \\"bundledDependencies\\": {
    \\"express\\": \\"^1.0.0\\",
    \\"lodash\\": \\"^2.0.0\\"
  },
  \\"optionalDependencies\\": {
    \\"bluebird\\": \\"^1.0.0\\",
    \\"fibers\\": \\"^2.0.0\\"
  },
  \\"peerDependencies\\": {
    \\"vue\\": \\"^1.0.0\\",
    \\"vue-template-compiler\\": \\"^2.0.0\\"
  },
  \\"publishConfig\\": {
    \\"access\\": \\"public\\",
    \\"registry\\": \\"https://internal.host.com/registry\\"
  },
  \\"eslintConfig\\": {
    \\"extends\\": \\"standard\\"
  },
  \\"jest\\": {
    \\"bail\\": 1,
    \\"verbose\\": true
  }
}
{
  \\"name\\": \\"full\\",
  \\"private\\": \\"true\\",
  \\"version\\": \\"1.2.3\\",
  \\"description\\": \\"Package description\\",
  \\"keywords\\": [
    \\"a\\",
    \\"b\\",
    \\"c\\"
  ],
  \\"engines\\": {
    \\"node\\": \\"12.0.0\\",
    \\"npm\\": \\"6.0.0\\"
  },
  \\"type\\": \\"module\\",
  \\"main\\": \\"dist/index.js\\",
  \\"module\\": \\"dist/index.mjs\\",
  \\"svelte\\": \\"src/index.js\\",
  \\"bin\\": \\"bin.js\\",
  \\"files\\": [
    \\"src\\",
    \\"test\\"
  ],
  \\"homepage\\": \\"https://host.com/test/repo\\",
  \\"bugs\\": \\"https://host.com/test/repo/issues\\",
  \\"repository\\": {
    \\"type\\": \\"git\\",
    \\"url\\": \\"git@host.com:test/repo.git\\"
  },
  \\"author\\": \\"Barney Rubble <b@rubble.com>\\",
  \\"contributors\\": [
    {
      \\"name\\": \\"Jane Smith\\",
      \\"email\\": \\"jane.smith@host.com\\"
    },
    \\"Jane Smith <jane.smith@host.com>\\",
    {
      \\"name\\": \\"John Smith\\",
      \\"email\\": \\"john.smith@host.com\\"
    },
    \\"John Smith <john.smith@host.com>\\"
  ],
  \\"license\\": \\"MIT\\",
  \\"scripts\\": {
    \\"build\\": \\"./build.sh\\",
    \\"lint\\": \\"./lint.sh\\",
    \\"test\\": \\"./test.sh\\"
  },
  \\"dependencies\\": {
    \\"express\\": \\"^1.0.0\\",
    \\"lodash\\": \\"^2.0.0\\"
  },
  \\"devDependencies\\": {
    \\"eslint\\": \\"^1.0.0\\",
    \\"jest\\": \\"^2.0.0\\"
  },
  \\"bundledDependencies\\": {
    \\"express\\": \\"^1.0.0\\",
    \\"lodash\\": \\"^2.0.0\\"
  },
  \\"optionalDependencies\\": {
    \\"bluebird\\": \\"^1.0.0\\",
    \\"fibers\\": \\"^2.0.0\\"
  },
  \\"peerDependencies\\": {
    \\"vue\\": \\"^1.0.0\\",
    \\"vue-template-compiler\\": \\"^2.0.0\\"
  },
  \\"publishConfig\\": {
    \\"access\\": \\"public\\",
    \\"registry\\": \\"https://internal.host.com/registry\\"
  },
  \\"eslintConfig\\": {
    \\"extends\\": \\"standard\\"
  },
  \\"jest\\": {
    \\"bail\\": 1,
    \\"verbose\\": true
  }
}
"
`;

exports[`$ nice-package-json prints a formatted package.json from the current directory 1`] = `
"{
  \\"name\\": \\"full\\",
  \\"private\\": \\"true\\",
  \\"version\\": \\"1.2.3\\",
  \\"description\\": \\"Package description\\",
  \\"keywords\\": [
    \\"a\\",
    \\"b\\",
    \\"c\\"
  ],
  \\"engines\\": {
    \\"node\\": \\"12.0.0\\",
    \\"npm\\": \\"6.0.0\\"
  },
  \\"type\\": \\"module\\",
  \\"main\\": \\"dist/index.js\\",
  \\"module\\": \\"dist/index.mjs\\",
  \\"svelte\\": \\"src/index.js\\",
  \\"bin\\": \\"bin.js\\",
  \\"files\\": [
    \\"src\\",
    \\"test\\"
  ],
  \\"homepage\\": \\"https://host.com/test/repo\\",
  \\"bugs\\": \\"https://host.com/test/repo/issues\\",
  \\"repository\\": {
    \\"type\\": \\"git\\",
    \\"url\\": \\"git@host.com:test/repo.git\\"
  },
  \\"author\\": \\"Barney Rubble <b@rubble.com>\\",
  \\"contributors\\": [
    {
      \\"name\\": \\"Jane Smith\\",
      \\"email\\": \\"jane.smith@host.com\\"
    },
    \\"Jane Smith <jane.smith@host.com>\\",
    {
      \\"name\\": \\"John Smith\\",
      \\"email\\": \\"john.smith@host.com\\"
    },
    \\"John Smith <john.smith@host.com>\\"
  ],
  \\"license\\": \\"MIT\\",
  \\"scripts\\": {
    \\"build\\": \\"./build.sh\\",
    \\"lint\\": \\"./lint.sh\\",
    \\"test\\": \\"./test.sh\\"
  },
  \\"dependencies\\": {
    \\"express\\": \\"^1.0.0\\",
    \\"lodash\\": \\"^2.0.0\\"
  },
  \\"devDependencies\\": {
    \\"eslint\\": \\"^1.0.0\\",
    \\"jest\\": \\"^2.0.0\\"
  },
  \\"bundledDependencies\\": {
    \\"express\\": \\"^1.0.0\\",
    \\"lodash\\": \\"^2.0.0\\"
  },
  \\"optionalDependencies\\": {
    \\"bluebird\\": \\"^1.0.0\\",
    \\"fibers\\": \\"^2.0.0\\"
  },
  \\"peerDependencies\\": {
    \\"vue\\": \\"^1.0.0\\",
    \\"vue-template-compiler\\": \\"^2.0.0\\"
  },
  \\"publishConfig\\": {
    \\"access\\": \\"public\\",
    \\"registry\\": \\"https://internal.host.com/registry\\"
  },
  \\"eslintConfig\\": {
    \\"extends\\": \\"standard\\"
  },
  \\"jest\\": {
    \\"bail\\": 1,
    \\"verbose\\": true
  }
}
"
`;
